src/ddmd/mars.d

```diff
@@ -245,6 +245,54 @@ extern (C++) void genCmain(Scope* sc)
     rootHasMain = sc._module;
 }
 
+/**
+ * Copies data from a FILE* into a newly-allocated buffer.
+ *
+ * The buffer is always null-terminated.
+ */
+ubyte[] copyFile(FILE* fp)
+{
+    import ddmd.root.rmem : mem;
+
+    enum bufIncrement = 4096;
+    size_t pos = 0;
+    size_t sz = bufIncrement;
+
+    ubyte* buf = cast(ubyte*)mem.xmalloc(sz + 1); // +1 for sentinel
+    if (buf is null)
+    {
+        error(Loc(), "out of memory");
+        fatal();
+    }
+
+    assert(sz > pos);
+    size_t rlen;
+    pos += rlen = fread(buf + pos, 1, sz - pos, fp);
+    while (!ferror(fp) && !feof(fp))
+    {
+        if (pos == sz)
+        {
+            sz += bufIncrement;
```

Comment:
Reallocating every 4K bytes makes for a very slow read. `expression.d` is 500K, so that makes for 125 copies and reallocations. I'd probably make it at least 128K.